provider/provider.go

```diff
@@ -113,3 +113,48 @@ func pinSet(ctx context.Context, pinning pin.Pinner, fetchConfig fetcher.Factory
 
 	return set, nil
 }
+
+func NewPrioritizedProvider(streams ...KeyChanFunc) KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		outCh := make(chan cid.Cid)
+
+		go func() {
+			defer close(outCh)
+			has := map[string]bool{}
```

Comment:
:thought_balloon: since we use it as set-like data structure, and care only about existence of key, and not value, we can avoid allocating bool values:
```suggestion
			has := map[string]struct{}
```


and then:

```diff
- has[str] = true
+ has[str] = struct{}{}
```